modules/frontend/pipeline/async_sharding.go

```diff
@@ -0,0 +1,104 @@
+package pipeline
+
+import (
+	"net/http"
+	"sync"
+
+	"github.com/grafana/tempo/pkg/boundedwaitgroup"
+)
+
+type waitGroup interface {
+	Add(int)
+	Done()
+	Wait()
+}
+
+// NewAsyncSharder creates a new AsyncResponse that shards requests to the next AsyncRoundTripper[*http.Response]. It creates one
+// goroutine per concurrent request.
+func NewAsyncSharder(concurrent int, reqFn func(i int) *http.Request, next AsyncRoundTripper[*http.Response]) Responses[*http.Response] {
+	var wg waitGroup
+	if concurrent <= 0 {
+		wg = &sync.WaitGroup{}
+	} else {
+		bwg := boundedwaitgroup.New(uint(concurrent))
+		wg = &bwg
+	}
+	asyncResp := newAsyncResponse()
+
+	go func() {
+		defer asyncResp.done()
+
+		reqIdx := 0
+		for {
+			req := reqFn(reqIdx)
+			reqIdx++
+
+			// else check for a request to pass down the pipeline
+			if req == nil {
```

Comment:
What do you think about the caller passing the expected number of requests?  It looks like the only reference is for multi-tenant where the number is known.